File: src/scripts/figure/festival_guy.js
log_info("akhenaten: figure festival_guy started")

[es=(figure_festival_guy, setup_phrase)]
function figure_festival_guy_setup_phrase(ev) {
    var f = city.get_figure(ev.fid)
    if (!f || !f.valid) {
        return
    }

    var home = f.home
    if (!home || !home.valid) {
        figure_apply_phrase(f, "empty")
        return
    }

    switch (home.type) {
    case BUILDING_TEMPLE_OSIRIS:
    case BUILDING_TEMPLE_COMPLEX_OSIRIS:
        figure_apply_phrase(f, "osiris_god_love_festival")
        break
    case BUILDING_TEMPLE_RA:
    case BUILDING_TEMPLE_COMPLEX_RA:
        figure_apply_phrase(f, "ra_god_love_festival")
        break
    case BUILDING_TEMPLE_PTAH:
    case BUILDING_TEMPLE_COMPLEX_PTAH:
        figure_apply_phrase(f, "ptah_god_love_festival")
        break
    case BUILDING_TEMPLE_SETH:
    case BUILDING_TEMPLE_COMPLEX_SETH:
        figure_apply_phrase(f, "seth_god_love_festival")
        break
    case BUILDING_TEMPLE_BAST:
    case BUILDING_TEMPLE_COMPLEX_BAST:
        figure_apply_phrase(f, "bast_god_love_festival")
        break
    case BUILDING_JUGGLER_SCHOOL:
        figure_apply_phrase(f, "juggler_i_like_festivals")
        break
    case BUILDING_CONSERVATORY:
        figure_apply_phrase(f, "musician_i_like_festivals")
        break
    case BUILDING_DANCE_SCHOOL:
        figure_apply_phrase(f, "dancer_i_like_festivals")
        break
    default:
        figure_apply_phrase(f, "empty")
        break
    }
}

File: src/figuretype/festival_guy.cpp
#include "figuretype/festival_guy.h"

#include "core/random.h"
#include "city/city.h"
#include "city/buildings.h"
#include "js/js_game.h"

REPLICATE_STATIC_PARAMS_FROM_CONFIG(figure_festival_guy);

void figure_festival_guy::update_animation() {
    building* temple = home();
    if (!temple->is_valid()) {
        return;
    }

    xstring animkey = {};
    switch (temple->type) {
    case BUILDING_TEMPLE_OSIRIS:
    case BUILDING_TEMPLE_COMPLEX_OSIRIS:
        animkey = "priest_osiris_walk";
        break;

    case BUILDING_TEMPLE_RA:
    case BUILDING_TEMPLE_COMPLEX_RA:
        animkey = "priest_ra_walk";
        break;

    case BUILDING_TEMPLE_PTAH:
    case BUILDING_TEMPLE_COMPLEX_PTAH:
        animkey = "priest_ptah_walk";
        break;

    case BUILDING_TEMPLE_SETH:
    case BUILDING_TEMPLE_COMPLEX_SETH:
        animkey = "priest_seth_walk";
        break;

    case BUILDING_TEMPLE_BAST:
    case BUILDING_TEMPLE_COMPLEX_BAST:
        animkey = "priest_bast_walk";
        break;

    case BUILDING_JUGGLER_SCHOOL:
        animkey = "juggler_walk";
        break;
    case BUILDING_CONSERVATORY:
        animkey = "musician_walk";
        break;
    case BUILDING_DANCE_SCHOOL:
        animkey = "dancer_walk";
        break;

    default:
        return;
    }

    image_set_animation(animkey);
}

void figure_festival_guy::figure_action() {
    switch (action_state()) {
    case ACTION_10_FESTIVAL_GUY_CREATED:
        base.animctx.frame = 0;
        if (--base.wait_ticks <= 0) {
            advance_action(ACTION_11_FESTIVAL_GUY_GOTO_SQUARE);
        }
        break;

    case ACTION_11_FESTIVAL_GUY_GOTO_SQUARE:
        if (do_goto(base.destination_tile, TERRAIN_USAGE_ANY, ACTION_12_FESTIVAL_GUY_DANCE)) {
            base.wait_ticks = rand() % 20;
        }
        break;

    case ACTION_12_FESTIVAL_GUY_DANCE:
        if (--base.wait_ticks <= 0) {
            advance_action(ACTION_13_FESTIVAL_GUY_GOTO_HOME);
        }
        break;

    case ACTION_13_FESTIVAL_GUY_GOTO_HOME:
        do_gotobuilding(home());
        break;
    }
}

void figure_festival_guy::before_poof() {
    ; //
}